### PR TITLE
Bug 2021322: Azure Marketplace Purchase Plan Info, Take Two

### PR DESCRIPTION
**Pull request by patrickdillon**
#237 populated purchase plan info whenever a marketplace image is used, but that implementation broke when the image was free and the plan was nil. Rather than the approach of #237 which inferred the purchase plan from the specified image, this implementation uses the virtual machine image api to get the image, and then reads the plan from that image, including if the image is free and has a nil plan.

I can drop b991f9761c0bfebd4e16b14bbfc5b1128562314f, which reverts #237, once #239 merges.

I wasn't sure what the best way to introduce this new client for VM Images was, so I put this together and figured we could hash it out in review. This is more substantial than the previous PR so I'm definitely open to any feedback or changes of plan.

### CODE REVIEW

File: pkg/cloud/azure/services/virtualmachines/service.go [hunk 1]
```diff
@@ -25,26 +25,36 @@ import (
 
 // Service provides operations on resource groups
 type Service struct {
-	Client compute.VirtualMachinesClient
-	Scope  *actuators.MachineScope
+	Client      compute.VirtualMachinesClient
+	ImageClient compute.VirtualMachineImagesClient
```
Comment (JoelSpeed) [inline pkg/cloud/azure/services/virtualmachines/service.go:29]: Nit, maybe `ImagesClient` to be consistent with the interface name?

**Review by JoelSpeed — COMMENTED** [review]
Would it be possible to add a unit test so we can see that this does/doesn't add the plan details when using a marketplace image thats paid for and free respectively?

**Comment by patrickdillon** [comment]
> Would it be possible to add a unit test so we can see that this does/doesn't add the plan details when using a marketplace image thats paid for and free respectively?

+1. The only way I see how to do this is make the images client an interface. I think that would still be pretty clean. I will throw that together and we can discuss from there.

**Comment by patrickdillon** [comment]
I added an interface and mocks for testing. 

I noticed there are higher-level mocks and interfaces here: https://github.com/openshift/cluster-api-provider-azure/tree/master/pkg/cloud/azure

It did not seem right to add the VM testing mocks there, but I'm open for any suggestions.

**Comment by patrickdillon** [comment]
>  we need to work out what went wrong with the pre-merge verification

There error is 
```
machine: compute.VirtualMachineImagesClient#Get: Failure responding to request:
StatusCode=400 -- Original Error: autorest/azure: Service returned an error.
Status=400 Code="InvalidParameter" Message="The value of parameter version is
invalid." Target="version"'
```

I get this error when passing `"latest"` as the `version` argument to `VirtualMachineImagesClient.Get`

So I think we need to figure out how to handle "latest" version.

**Comment by patrickdillon** [comment]
/hold

We're going to try a different approach with adding the parameters to the machine spec.

**Comment by patrickdillon** [comment]
> /hold
> 
> We're going to try a different approach with adding the parameters to the machine spec.

The reason we are doing this is because it is valid for users to fill out the machine spec with a `latest` `version`, but if users do that, the current approach for this PR would then need to determine what is the latest version of the image being specified, then use that particular image to query.

Adding this much logic makes this approach less viable, than following the upstream approach of letting clients specify these fields through the machine spec.

**Comment by patrickdillon** [comment]
/close
in favor of take three (!) #242

**Pull request closed without merging**